src/inventorymanager.cpp

```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
+	else if (ma.to_inv.type == InventoryLocation::NODEMETA)
+		can_put_count = sa->nodemeta_inventory_AllowPut(ma, item, player);
+	else if (ma.to_inv.type == InventoryLocation::PLAYER)
+		can_put_count = sa->player_inventory_AllowPut(ma, item, player);
 	else
 		assert(false);
-	return dst_can_put_count;
+	return can_put_count;
 }
 
-int IMoveAction::allowTake(const ItemStack &src_item, ServerActiveObject *player) const
+static int allowTake(const MoveAction& ma, const ItemStack &item, ServerActiveObject *player)
 {
 	ServerScripting *sa = PLAYER_TO_SA(player);
-	int src_can_take_count = 0xffff;
-	if (from_inv.type == InventoryLocation::DETACHED)
-		src_can_take_count = sa->detached_inventory_AllowTake(*this, src_item, player);
-	else if (from_inv.type == InventoryLocation::NODEMETA)
-		src_can_take_count = sa->nodemeta_inventory_AllowTake(*this, src_item, player);
-	else if (from_inv.type == InventoryLocation::PLAYER)
-		src_can_take_count = sa->player_inventory_AllowTake(*this, src_item, player);
+	int can_take_count = 0xffff;
+	if (ma.from_inv.type == InventoryLocation::DETACHED)
+		can_take_count = sa->detached_inventory_AllowTake(ma, item, player);
+	else if (ma.from_inv.type == InventoryLocation::NODEMETA)
+		can_take_count = sa->nodemeta_inventory_AllowTake(ma, item, player);
+	else if (ma.from_inv.type == InventoryLocation::PLAYER)
+		can_take_count = sa->player_inventory_AllowTake(ma, item, player);
 	else
 		assert(false);
-	return src_can_take_count;
+	return can_take_count;
 }
 
-int IMoveAction::allowMove(int try_take_count, ServerActiveObject *player) const
+static int allowMove(const MoveAction& ma, int try_take_count, ServerActiveObject *player)
 {
 	ServerScripting *sa = PLAYER_TO_SA(player);
-	int src_can_take_count = 0xffff;
-	if (from_inv.type == InventoryLocation::DETACHED)
-		src_can_take_count = sa->detached_inventory_AllowMove(*this, try_take_count, player);
-	else if (from_inv.type == InventoryLocation::NODEMETA)
-		src_can_take_count = sa->nodemeta_inventory_AllowMove(*this, try_take_count, player);
-	else if (from_inv.type == InventoryLocation::PLAYER)
-		src_can_take_count = sa->player_inventory_AllowMove(*this, try_take_count, player);
+	int can_take_count = 0xffff;
+	if (ma.from_inv.type == InventoryLocation::DETACHED)
+		can_take_count = sa->detached_inventory_AllowMove(ma, try_take_count, player);
+	else if (ma.from_inv.type == InventoryLocation::NODEMETA)
+		can_take_count = sa->nodemeta_inventory_AllowMove(ma, try_take_count, player);
+	else if (ma.from_inv.type == InventoryLocation::PLAYER)
+		can_take_count = sa->player_inventory_AllowMove(ma, try_take_count, player);
 	else
 		assert(false);
-	return src_can_take_count;
+	return can_take_count;
+}
+
+static std::pair<s16, bool> move(const MoveAction& ma, InventoryList *list_from, InventoryList *list_to,
+	u16 count, bool ignore_rollback, InventoryManager *mgr, ServerActiveObject *player, IGameDef *gamedef)
+{
```

Comment:
```suggestion
// returns {moved_count, can_swap}
static std::pair<s16, bool> move(const MoveAction& ma,
	InventoryList *list_from, InventoryList *list_to, u16 count, bool ignore_rollback,
	InventoryManager *mgr, ServerActiveObject *player, IGameDef *gamedef)
{
```

Reply:
Why is this not a member function of `MoveAction`? That object holds the inventory list data and move count. It would shorten this function call quite a bit.

Reply:
Fixed